Resolve null-dereference and enable Linux containers

davidlt opened the pull request:
Notice: commit messages contains more information.

CMSSW fails on at least one site where Linux containers are used (probably instead of virtual machines). You cannot trust `/etc/mtab` content, go directly for `/proc/self/mounts`.

Same issue happens if CMSSW is compiled with ASan, in this case `realpath` was not doing a correct job for "`.`". Could be a bug in ASan interceptor.

Commit:
Avoid null-dereference in LocalFileSystem::findMount

After initial for-loop best might be 0x0. The patch avoids
NULL-dereference issue, which causes segfaults.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>
Reported-by: Sébastien Brochet <sebastien.brochet@cern.ch>

cmsbuild commented:
A new Pull Request was created by @davidlt  for CMSSW_7_5_X.

Resolve null-dereference and enable Linux containers

It involves the following packages:

Utilities/StorageFactory


@cmsbuild, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

davidlt commented:
This also would be needed in 7_4_X.

cmsbuild commented:
Pull request #9541 was updated. @cmsbuild, @Dr15Jones can you please check and sign again.

davidlt commented:
Tests are scheduled.

Dr15Jones commented on Utilities/StorageFactory/src/LocalFileSystem.cc:
```diff
@@ -212,6 +212,7 @@ LocalFileSystem::initFSInfo(void *arg)
 int
 LocalFileSystem::initFSList(void)
 {
+#define _PATH_MOUNTED_LINUX "/proc/self/mounts"
```
Change this to a ```const char * const PATH_MOUNTED_LINUX``` and declaring it only in the block it is needed.

Dr15Jones commented:
@bbockelm @dan131riley do you guys have comments from a systems perspective?

davidlt commented:
Modified. You can check the diff again.

cmsbuild commented:
Pull request #9541 was updated. @cmsbuild, @Dr15Jones can you please check and sign again.

Commit:
Use /proc/self/mounts instead of /etc/mtab

According to glibc documentation one should use `_PATH_MOUNTED`, which
by default should point to `/etc/mtab`. This file is maintained by
`mount` and `umount` programs.

This is wrong if we are running in a sandboxed everinment, e.g. Linux
containers (OpenVZ, Docker, Rocket, LXC/LXD) or chroot/proot. Then this
files comes from alternative rootfs and does not reflect actual mount
points.

On modern systems (incl. CentOS 7 and Fedora) `/etc/mtab` points to
`/proc/self/mounts`.

The patch by default reads `/proc/self/mounts` on Linux systems.

By default Docker will create a `/etc/mtab` as symlink to `/proc/mounts`
(which points to `/proc/self/mounts`) on any container.

OpenVZ documentation also request chaging `/etc/mtab` to a symlink.

Same in Linux From Scratch guide. Also reminds that `/` could be mounted
as read-only, thus `/etc/mtab` could be stale.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>
Reported-by: Sébastien Brochet <sebastien.brochet@cern.ch>

bbockelm commented:
This looks appropriate to me.  Use of /etc/mtab is a bit historical (this is old code) anyway - I think /proc/self/mounts is really the correct way going forward.

cmsbuild commented:
Pull request #9541 was updated. @cmsbuild, @Dr15Jones can you please check and sign again.

blinkseb commented:
Hi there,

I can confirm that this patch fix the crash I observed on our remote UIs where /etc/mtab is not in sync with /proc/self/mounts.

Thanks!

cmsbuild commented:
+1
Tested at: 19110ea00f2a5968f3becf398cf4ca824eee020c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9541/4085/summary.html

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9541/4085/summary.html

There are some workflows for which there are errors in the baseline:
 25202.0 step 2
50202.0 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
